From 04fa05693118f7258f6d72ea95ceb1d34b126e17 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 18:59:56 -0400
Subject: [fs] btrfs: optimize to remove unnecessary removal with ulist reallocation

Message-id: <1407179109-13714-110-git-send-email-sandeen@redhat.com>
Patchwork-id: 87068
O-Subject: [RHEL7.1 PATCH 109/422] Btrfs: optimize to remove unnecessary removal with ulist reallocation
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Wang Shilong <wangsl.fnst@cn.fujitsu.com>

commit f499e40fd97698a1c48d188279647009b21905fe
Author: Wang Shilong <wangsl.fnst@cn.fujitsu.com>
Date:   Fri Jan 10 21:25:46 2014 +0800

    Btrfs: optimize to remove unnecessary removal with ulist reallocation

    Here we are not going to free memory, no need to remove every node
    one by one, just init root node here is ok.

    Cc:  Liu Bo <bo.li.liu@oracle.com>
    Signed-off-by: Wang Shilong <wangsl.fnst@cn.fujitsu.com>
    Signed-off-by: Josef Bacik <jbacik@fb.com>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/ulist.c b/fs/btrfs/ulist.c
index b0a523b..35f5de9 100644
--- a/fs/btrfs/ulist.c
+++ b/fs/btrfs/ulist.c
@@ -207,9 +207,6 @@ int ulist_add_merge(struct ulist *ulist, u64 val, u64 aux,
   void *old = NULL;
   int i;
 
-  for (i = 0; i < ulist->nnodes; i++)
-   rb_erase(&ulist->nodes[i].rb_node, &ulist->root);
-
   /*
    * if nodes_alloced == ULIST_SIZE no memory has been allocated
    * yet, so pass NULL to krealloc
@@ -234,6 +231,7 @@ int ulist_add_merge(struct ulist *ulist, u64 val, u64 aux,
    * pointers, so we have to do it ourselves.  Otherwise we may
    * be bitten by crashes.
    */
+  ulist->root = RB_ROOT;
   for (i = 0; i < ulist->nnodes; i++) {
    ret = ulist_rbtree_insert(ulist, &ulist->nodes[i]);
    if (ret < 0)
-- 
1.7.1